src/Functions/array/arrayIndex.h

```diff
@@ -724,227 +688,153 @@ class FunctionArrayIndex : public IFunction
         }
-        else // LowCardinality(T) and U, T not Nullable
-        {
-            if (arguments[1].column->isNullable())
-                return nullptr;
 
-            if (const auto* const arg_lc = checkAndGetColumn<ColumnLowCardinality>(arguments[1].column.get());
-                arg_lc && arg_lc->isNullable())
-                return nullptr;
+        Impl::Main<ConcreteAction, true>::vector(
+            left_lc->getIndexes(),
+            col_array->getOffsets(),
+            index, /** Assuming LowCardinality has index of NULL always as zero. */
+            col_result->getData(),
+            nullptr,
+            nullptr);
 
-            // LowCardinality(T) and U (possibly LowCardinality(V))
+        return col_result;
+    }
 
-            const ColumnPtr left_casted = col_lc->convertToFullColumnIfLowCardinality();
-            const ColumnPtr right_casted = arguments[1].column->convertToFullColumnIfLowCardinality();
+    ColumnPtr executeMap(const ColumnsWithTypeAndName & arguments, const DataTypePtr & result_type) const
+    {
+        if constexpr (!std::is_same_v<ConcreteAction, HasAction>)
+            return nullptr;
 
-            ExecutionData data =
-            {
-                *left_casted.get(), *right_casted.get(), col_array->getOffsets(),
-                nullptr, {null_map_data, null_map_item}
-            };
+        if (!isMap(arguments[0].type))
+            return nullptr;
 
-            if (dispatchConvertedLowCardinalityColumns(data))
-                return data.result_column;
-        }
+        auto non_const_map_column = arguments[0].column->convertToFullColumnIfConst();
 
-        return nullptr;
-    }
+        const auto & map_column = assert_cast<const ColumnMap &>(*non_const_map_column);
+        const auto & map_array_column = map_column.getNestedColumn();
+        auto offsets = map_array_column.getOffsetsPtr();
+        auto keys = map_column.getNestedData().getColumnPtr(0);
+        auto array_column = ColumnArray::create(keys, offsets);
 
-    static bool dispatchConvertedLowCardinalityColumns(ExecutionData & data)
-    {
-        if (data.left.isNumeric() && data.right.isNumeric()) // ColumnArrays
-            return executeIntegral<INTEGRAL_TPL_PACK>(data);
-
-        if (checkAndGetColumn<ColumnString>(&data.left))
-            return executeStringImpl(data);
+        const auto & type_map = assert_cast<const DataTypeMap &>(*arguments[0].type);
+        auto array_type = std::make_shared<DataTypeArray>(type_map.getKeyType());
 
-        Impl::Main<ConcreteAction, true>::vector(
-            data.left,
-            data.offsets, data.right,
-            data.result->getData(),
-            data.maps.first, data.maps.second);
+        auto arguments_copy = arguments;
+        arguments_copy[0].column = std::move(array_column);
+        arguments_copy[0].type = std::move(array_type);
+        arguments_copy[0].name = arguments[0].name;
 
-        data.moveResult();
-        return true;
```

Comment:
The problem was here: we executed "generic" implementation without cast of columns to common type. Also this code looked strange.